pkg/typesystem/typesystem_test.go

```diff
@@ -12,6 +12,490 @@ import (
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`this_has_entrypoints_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_user_wildcard`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document:*]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_userset`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type org
+				relations
+					define member: [user]
+			type folder
+				relations
+					define parent: [org#member]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_first`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define parent: [user, folder#parent]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_second`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define editor: [user]
+					define parent: [folder#parent, folder#editor]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		// TODO fix
+		//`this_has_no_entrypoints_because_type_unknown_is_not_defined`: {
+		//	model: `
+		//	model
+		//		schema 1.1
+		//	type folder
+		//		relations
+		//			define parent: [unknown]`,
+		//	inputType:     "folder",
+		//	inputRelation: "parent",
+		//	expectError:   "undefined type 'unknown'",
+		//},
+		`this_has_no_entrypoints`: {
+			model: `
+			model
+				schema 1.1
+			type folder
+				relations
+					define parent: [folder#parent]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{false, false},
+		},
+		`computed_relation_has_entrypoint_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type document
+				relations
+					define editor: [user]
+					define viewer: editor`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`computed_relation_has_entrypoint_through_userset`: {
+			model: `
+			model
+			  schema 1.1
+			type user
+			type org
+			  relations
+				define member: [user]
+			type folder
+			  relations
+				define a2: [org#member]
+				define a1: a2`,
+			inputType:     "folder",
+			inputRelation: "a1",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`computed_relation_has_no_entrypoints_because_no_terminal_node`: {
```

Comment:
```suggestion
		`computed_relation_has_no_entrypoints_because_no_direct_relationships`: {
```

or 
`computed_relation_has_no_entrypoints_because_no_terminal_relationships`

I know what you mean by node, but I think relationship is more aptly used here.